probe-rs/src/bin/probe-rs/cmd/test.rs

```diff
@@ -0,0 +1,472 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use anyhow::{anyhow, bail, Result};
+use libtest_mimic::{Failed, Trial};
+use probe_rs::{
+    BreakpointCause, Core, CoreStatus, HaltReason, Lister, MemoryInterface, SemihostingCommand,
+};
+use probe_rs_target::MemoryRegion;
+use signal_hook::consts::signal;
+use time::UtcOffset;
+
+use crate::util::common_options::{BinaryDownloadOptions, ProbeOptions, RunOptions};
+use crate::util::flash::{build_loader, run_flash_download};
+use crate::util::rtt::{self, poll_rtt, try_attach_to_rtt};
+use crate::util::stack_trace::print_stacktrace;
+use crate::FormatOptions;
+
+#[derive(clap::Parser)]
+pub struct Cmd {
+    #[clap(flatten)]
+    pub(crate) probe_options: ProbeOptions,
+
+    #[clap(flatten)]
+    pub(crate) download_options: BinaryDownloadOptions,
+
+    #[clap(flatten)]
+    pub(crate) run_options: RunOptions,
+
+    /// Whether to erase the entire chip before downloading
+    #[clap(long)]
+    pub(crate) chip_erase: bool,
+
+    #[clap(flatten)]
+    pub(crate) format_options: FormatOptions,
+
+    /// <elf> and the remaining arguments for the test runner (list tests, filter tests etc). Run `probe-rs test -- <elf> --help` for more information.
+    #[clap(last(true))]
+    pub(crate) libtest_args: Vec<String>,
+}
+
+struct Runner<'c> {
+    core: Core<'c>,
+    timestamp_offset: UtcOffset,
+    path: String,
+    always_print_stacktrace: bool,
+    no_location: bool,
+    log_format: Option<String>,
+    memory_map: Vec<MemoryRegion>,
+    rtt_scan_regions: Vec<Range<u64>>,
+}
+
+impl Cmd {
+    pub fn run(
+        self,
+        lister: &Lister,
+        run_download: bool,
+        timestamp_offset: UtcOffset,
+    ) -> Result<()> {
+        let path = self.libtest_args[0].clone();
+        let libtest_args = libtest_mimic::Arguments::from_iter(self.libtest_args);
+
+        let (mut session, probe_options) = self.probe_options.simple_attach(lister)?;
+
+        if run_download {
+            let path = Path::new(&path);
+            let loader = build_loader(&mut session, path, self.format_options)?;
+            run_flash_download(
+                &mut session,
+                &path,
+                &self.download_options,
+                &probe_options,
+                loader,
+                self.chip_erase,
+            )?;
+            // reset the core to leave it in a consistent state after flashing
+            session
+                .core(0)?
+                .reset_and_halt(Duration::from_millis(100))?;
+        }
+
+        let memory_map = session.target().memory_map.clone();
+        let rtt_scan_regions = match self.run_options.rtt_scan_memory {
+            true => session.target().rtt_scan_regions.clone(),
+            false => Vec::new(),
+        };
+
+        let mut core = session.core(0)?;
+
+        // TODO: We need to enable_vector_catch after every chip reset, or is this persistent?
```

Comment:
@Yatekii 
can you help here?

Reply:
Yes, you need to enable it after every reset afaik :)